### PR TITLE
Added macro generation in MLF export

### PR DESCRIPTION
**Pull request by fPecc**
For description see this [forum topic](https://discuss.tvm.apache.org/t/exporting-generated-functions-input-and-output-sizes-in-the-model-library-format/13532).

Tagged reviewers: @areusch (don't know who else can review microTVM-related code)

### CODE REVIEW

**Comment by leandron** [comment]
cc @Mousius @ashutosh-arm @grant-arm

**Comment by leandron** [comment]
@fPecc from what I understood from reading your Discuss post, this is a non-breaking change, right?

Apart from that, I think the problem is quite restrict and it could've been summarised/discussed in the PR, rather than in a separate location, but that's just a minor suggestion.

File: src/target/source/interface_c.cc [hunk 1]
```diff
@@ -46,20 +46,46 @@ class InterfaceCNode : public runtime::ModuleNode {
...
 
   std::string GetSource(const std::string& format) final {
     std::stringstream code;
 
     EmitUpperHeaderGuard(code);
+
+    // Emit macros for input sizes
+    for (auto const& it : input_sizes_) {
+      auto input_name = std::string(it.first);
+      std::replace(input_name.begin(), input_name.end(), ':', '_');
```
Comment (areusch) [inline src/target/source/interface_c.cc:70]: can you consolidate this logic so we can reuse it?

File: src/target/source/interface_c.cc [hunk 2]
```diff
@@ -46,20 +46,46 @@ class InterfaceCNode : public runtime::ModuleNode {
...
+      std::replace(input_name.begin(), input_name.end(), '+', '_');
+      std::replace(input_name.begin(), input_name.end(), '-', '_');
+      auto input_size = it.second->value;
+      EmitIntegerValueMacro(code, "Input tensor " + input_name + " size (in bytes)", input_name,
```
Comment (areusch) [inline src/target/source/interface_c.cc:74]: can we include `_SIZE` in the macro name like the other postfixes? also, we might want to consider this as a prefix after "tvmgen_modname_" to avoid potential collision (if for some reason a layer function has the same name).

**Review by areusch — COMMENTED** [review]
thanks @fPecc for this suggestion! I'm generally ok with this, my main ask would be whether or not we should consider other properties and how that relates to information we might export in metadata.json. Overall, I'm not too worried if we make breaking naming changes here in the future as TVM is still <1.0, but it would just add an item to our tech debt is all. I think being able to reference this info easily is useful though, so I'm not opposed to merging.

cc @alanmacd @mehrdadh @mkatanbaf @guberti @gromero

**Comment by fPecc** [comment]
Thanks for the feedback @areusch . Indeed, it shouldn't break, but there are still some tests that are not passing, so I am trying to fix that locally before continuing and commiting again.

File: src/target/source/interface_c.cc [hunk 3]
```diff
@@ -46,20 +46,46 @@ class InterfaceCNode : public runtime::ModuleNode {
...
     std::stringstream code;
 
     EmitUpperHeaderGuard(code);
+
+    // Emit macros for input sizes
+    for (auto const& it : input_sizes_) {
+      auto input_name = std::string(it.first);
+      std::replace(input_name.begin(), input_name.end(), ':', '_');
+      std::replace(input_name.begin(), input_name.end(), '+', '_');
+      std::replace(input_name.begin(), input_name.end(), '-', '_');
```
Comment (Mousius) [inline src/target/source/interface_c.cc:72]: We can use the existing name sanitisation logic here:

```suggestion
      std::string input_name = SanitizeName(it.first);
```

File: src/target/source/interface_c.cc [hunk 4]
```diff
@@ -46,20 +46,46 @@ class InterfaceCNode : public runtime::ModuleNode {
...
+      auto input_size = it.second->value;
```
Comment (Mousius) [inline src/target/source/interface_c.cc:73]: Where possible, we should avoid `auto` types:

```suggestion
      int input_size = it.second->value;
```

File: python/tvm/micro/model_library_format.py [hunk 5]
```diff
@@ -519,7 +570,7 @@ def _export_operator_model_library_format(mod: build_module.OperatorModule, temp
     """
     targets = []
     for target in mod.ir_module_by_target.keys():
-        if str(target.kind) not in ("llvm", "c"):
+        if str(target.kind) not in ("llvm", "c", "gemmini"):
```
Comment (Mousius) [inline python/tvm/micro/model_library_format.py:573]: This seems like an unrelated change?

File: src/target/source/interface_c.cc [hunk 6]
```diff
@@ -46,20 +46,46 @@ class InterfaceCNode : public runtime::ModuleNode {
...
+                            input_size);
+    }
+
+    // Emit macros for output sizes
+    for (auto const& it : output_sizes_) {
+      auto output_name = std::string(it.first);
+      std::replace(output_name.begin(), output_name.end(), ':', '_');
+      std::replace(output_name.begin(), output_name.end(), '+', '_');
+      std::replace(output_name.begin(), output_name.end(), '-', '_');
+      auto output_size = it.second->value;
```
Comment (Mousius) [inline src/target/source/interface_c.cc:84]:
```suggestion
      std::string output_name = SanitizeName(it.first);
      int output_size = it.second->value;
```

**Review by Mousius — CHANGES_REQUESTED** [review]

File: python/tvm/micro/model_library_format.py (hunk 5 above)
Comment (fPecc) [reply python/tvm/micro/model_library_format.py:573]: Yeah, sorry for that, will change it back.

Comment (fPecc) [reply python/tvm/micro/model_library_format.py]: added a new commit fixing this

File: src/target/source/interface_c.cc (hunk 2 above)
Comment (fPecc) [reply src/target/source/interface_c.cc]: just added it in a new commit

File: src/target/source/interface_c.cc (hunk 1 above)
Comment (fPecc) [reply src/target/source/interface_c.cc]: Added the usage of SanitizeName in order to keep a uniform "sanitization"

File: python/tvm/micro/model_library_format.py [hunk 7]
```diff
@@ -277,6 +294,37 @@ def _create_empty_entry(target_device_type):
...
+                if hasattr(output_type, "shape"):
+                    output_dict[i] = int(_shape_to_size(output_type.shape, output_type.dtype))
+                else:
+                    output_dict[i] = 0
+        else:
+            output_type = main_func_metadata.relay_primfuncs[target].ret_type
+            if hasattr(output_type, "shape"):
+                output_dict[0] = int(_shape_to_size(output_type.shape, output_type.dtype))
+            else:
+                output_dict[0] = 0
```
Comment (Mousius) [inline python/tvm/micro/model_library_format.py:325]: This is duplicated between input and output, we can replace it with a function:

```python
def _create_type_metadata(input_type):
    return {
        "size": int(
            _shape_to_size(input_type.shape, input_type.dtype)
        ),
        "dtype": str(input_type.dtype)
    }
```

File: python/tvm/micro/model_library_format.py [hunk 8]
```diff
@@ -277,6 +294,37 @@ def _create_empty_entry(target_device_type):
...
+                    _shape_to_size(input_param.checked_type.shape, input_param.checked_type.dtype)
+                )
+            else:
+                # TODO: maybe fill checked_type here?
+                input_dict[input_param.name_hint] = 0
+        target_main_entries[int(target.kind.device_type)]["inputs"] = input_dict
+
+        output_dict = {}
+        # For output, we dont have the name of the output, so we enumerate them
+        if isinstance(main_func_metadata.relay_primfuncs[target].ret_type, tvm.ir.type.TupleType):
```
Comment (Mousius) [inline python/tvm/micro/model_library_format.py:312]: Given tuples could be more heavily nested than this, we should use some recursive function to generate the outputs? We can re-purpose this function (which is no longer used): https://github.com/apache/tvm/blob/main/python/tvm/micro/model_library_format.py#L293

File: python/tvm/micro/model_library_format.py [hunk 9]
```diff
@@ -277,6 +294,37 @@ def _create_empty_entry(target_device_type):
             main_func_metadata.io_sizes[target]
         )
 
+        # Now, we also add the information about the size of each input and output of the main
+        # function (in bytes)
+        input_dict = {}
+        for input_param in main_func_metadata.relay_primfuncs[target].params:
+            if hasattr(input_param, "checked_type"):
```
Comment (Mousius) [inline python/tvm/micro/model_library_format.py:301]: I don't see a test case for when this is not set? How do we reproduce it?

File: tests/python/unittest/test_micro_model_library_format.py [hunk 10]
```diff
@@ -208,7 +208,12 @@ def @main(%a : Tensor[(1, 2), uint8], %b : Tensor[(1, 2), float32], %c : Tensor[
                 {
                     "constants_size_bytes": json_constants_size_bytes,
                     "device": 1,
+                    "inputs": {
+                        "a": 2,
+                        "b": 8,
+                    },
```
Comment (Mousius) [inline tests/python/unittest/test_micro_model_library_format.py:214]: For future proofing, it'd be good to have more than just the size? But also dtype here?

```suggestion
                    "inputs": {
                        "a": {"dtype": "uint8", "size_in_bytes": 2},
                        "b": {"dtype": "float32", "size_in_bytes": 8},
                    },
```

File: tests/python/unittest/test_micro_model_library_format.py [hunk 11]
```diff
@@ -208,7 +208,12 @@ def @main(%a : Tensor[(1, 2), uint8], %b : Tensor[(1, 2), float32], %c : Tensor[
...
                     "io_size_bytes": 18,
+                    "outputs": {"0": 8},
```
Comment (Mousius) [inline tests/python/unittest/test_micro_model_library_format.py:216]: This should match with the name of the output in the generated C (`output` for single, `outputX` for multiple, or the actual name if we can get to it)

```suggestion
                    "outputs": {"output": 8},
```

**Review by Mousius — CHANGES_REQUESTED** [review]

**Comment by mehrdadh** [comment]
@fPecc  Friendly reminder about this PR.

File: python/tvm/micro/model_library_format.py (hunk 7 above)
Comment (fPecc) [reply python/tvm/micro/model_library_format.py]: Applied the proposed change

File: python/tvm/micro/model_library_format.py (hunk 8 above)
Comment (fPecc) [reply python/tvm/micro/model_library_format.py:317]: Applied the proposed change

File: tests/python/unittest/test_micro_model_library_format.py (hunk 10 above)
Comment (fPecc) [reply tests/python/unittest/test_micro_model_library_format.py]: Added the check for the dtype

**Comment by fPecc** [comment]
@mehrdadh sorry for the delay, there it is, all tests fixed.

File: python/tvm/micro/model_library_format.py [hunk 12]
```diff
@@ -277,6 +301,32 @@ def _create_empty_entry(target_device_type):
...
+                if hasattr(output_type, "shape"):
+                    output_dict[f"output{i}"] = _create_type_metadata(output_type)
+                else:
+                    output_dict[f"output{i}"] = {"size": 0, "dtype": ""}
+        else:
+            output_type = main_func_metadata.relay_primfuncs[target].ret_type
+            if hasattr(output_type, "shape"):
+                output_dict["output"] = _create_type_metadata(output_type)
+            else:
+                output_dict["output"] = {"size": 0, "dtype": ""}
```
Comment (Mousius) [inline python/tvm/micro/model_library_format.py:327]: Is there a case where this happens? I can't see it in the test cases

**Review by Mousius — CHANGES_REQUESTED** [review]
Just one question but otherwise this looks good to me

File: python/tvm/micro/model_library_format.py (hunk 12 above)
Comment (fPecc) [reply python/tvm/micro/model_library_format.py]: You are right, there is no case this happens. I removed the check because it was redundant

**Comment by fPecc** [comment]
Hi @Mousius ,

I did the rebase, but I noticed there are still tests failing in CPU MINIMAL. They seem to be related to the CRT, but I cant find a direct relation with this change

**Comment by guberti** [comment]
@fPecc Are you sure you did a rebase, not a merge? I suspect `git rebase -i main` will fix the CRT tests.

**Review by Mousius — APPROVED** [review]

**Comment by Mousius** [comment]
Thanks for working through this one @fPecc! This is an awesome improvement and it has now landed 👍
